File: Python/InsufficientNodesinRoottoLeafPaths.py
import TreeNode

class Solution:
    def sufficientSubset(self, root: TreeNode, limit: int) -> TreeNode:
        res = self.dfs(root, 0, limit)
        if not res:
            return None
        return root


    def dfs(self, n: TreeNode, sum: int, limit: int) -> bool:
        if n == None:
            return False
        total = sum + n.val
        res = False
        if n.left == None and n.right == None:
            if total >= limit:
                res = True
            else:
                res = False
        else:
            l = self.dfs(n.left, total, limit)
            r = self.dfs(n.right, total, limit)
            if not l:
                n.left = None
            if not r:
                n.right = None
            res = l or r
        return res

File: Java/src/InsufficientNodesinRoottoLeafPaths.java
import java.util.HashSet;
import java.util.Stack;

public class InsufficientNodesinRoottoLeafPaths {

	public TreeNode sufficientSubset(TreeNode root, int limit) {
        HashSet<TreeNode> set = new HashSet<>();
        boolean res = dfs(root, 0, limit);
        
        if(!res) return null;
        return root;
    }

    private boolean dfs(TreeNode n, int sum, int limit) {
		if(n == null) return false;
        int total = sum + n.val;
        boolean res = false;
		if(n.left == null && n.right == null) {
			if(total >= limit) res = true;
			else res = false;
		}else {
			boolean l = dfs(n.left, total, limit);
            boolean r = dfs(n.right, total, limit);
            if(!l) n.left = null;
            if(!r) n.right = null;
			res = l || r;
        }
        
        return res;
	}
	
	private void helper(TreeNode n, HashSet<TreeNode> set) {
		if(n == null) return;
		if(!set.contains(n.left)) n.left = null;
		if(!set.contains(n.right)) n.right = null;
		
		helper(n.left, set);
		helper(n.right, set);
	}
	
	
	
}
